src/software/sensor_fusion/filter/ball_filter.cpp

```diff
@@ -0,0 +1,363 @@
+
+    auto x_vs_y_regression = getLinearRegressionLine(ball_detections);
+
+    // Linear regression cannot fit a vertical line. To get around this, we fit two lines,
+    // one with x and y swapped, so any vertical line becomes horizontal. Then we take the
+    // line of the two that fit the best.
+    boost::circular_buffer<BallDetection> swapped_ball_detections = ball_detections;
+    for (auto &detection : swapped_ball_detections)
+    {
+        detection.position = Point(detection.position.y(), detection.position.x());
+    }
+    auto y_vs_x_regression = getLinearRegressionLine(swapped_ball_detections);
+    // Because we swapped the coordinates of the input, we have to swap the coordinates of
+    // the output to get back to our expected coordinate space
+    y_vs_x_regression.regression_line.swapXY();
+
+    // We use the regression from above with the least error
+    Line regression_line =
+        x_vs_y_regression.regression_error < y_vs_x_regression.regression_error
+            ? x_vs_y_regression.regression_line
+            : y_vs_x_regression.regression_line;
+
+    // Take the position of the most recent ball position and project it onto the line of
+    // best fit. We do this because we assume the ball must be travelling along its
+    // velocity vector, and this allows us to return more stable position values since the
+    // line of best fit is less likely to fluctuate compared to the raw position of a ball
+    // detection
+    BallDetection latest_ball_detection = ball_detections.front();
+    Point filtered_ball_position =
+        closestPointOnLine(latest_ball_detection.position, regression_line);
+
+    std::optional<BallVelocityEstimate> velocity_estimate =
+        estimateBallVelocity(ball_detections, regression_line);
+    if (!velocity_estimate)
+    {
+        return std::nullopt;
+    }
+    else
+    {
+        // Project the velocity so it also lies along the line of best fit calculated
+        // above. Again, this gives more stable values because the line of best fit is
+        // more robust to fluctuations in ball position, so will not vary as much as using
+        // the "raw" ball velocity
+        auto velocity_direction_along_regression_line =
+            velocity_estimate->average_velocity.project(
+                regression_line.toNormalUnitVector().perpendicular());
+        Vector filtered_velocity = velocity_direction_along_regression_line.normalize(
+            velocity_estimate->average_velocity_magnitude);
+
+        return BallState(filtered_ball_position, filtered_velocity,
+                         latest_ball_detection.timestamp);
+    }
+}
+
+std::optional<Ball> BallFilter::getFilteredData(
+    const std::vector<BallDetection> &new_ball_detections, const Field &field)
+{
+    addNewDetectionsToBuffer(new_ball_detections, field);
+
+    if (ball_detection_buffer.size() >= 2)
+    {
+        std::optional<BallState> filtered_ball = estimateBallState(ball_detection_buffer);
```

Comment:
it feels like most of this function should be part of `estimateBallState`, so this function would look like:
```
addNewDetectionsToBuffer(new_ball_detections, field);

return estimateBallState(ball_detection_buffer);
```

Reply:
I don't disagree, but this is a "higher level" function that uses `estimateBallState` and returns a `Ball` rather than a `BallState`. I think `estimateBallState` or `getEstimatedBallState` is a better name for the public function that should be exposed, but not sure exactly what you're suggesting by combining theses functions.

Reply:
I can definitely refactor this class later (it probably needs it) but most filter changes are out of scope for this PR IMO? This is more about structure.

Reply:
Ok, could you file a ticket Mat? You probably have more context then I do for what needs to change.

Reply:
Created #1141

Reply:
is this resolved?

Reply:
yep, marking as such